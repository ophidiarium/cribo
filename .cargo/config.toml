[build]
rustflags = ["-Ctarget-cpu=native"]

[alias]
# Coverage commands using cargo-llvm-cov
coverage = "llvm-cov --html --open"
coverage-text = "llvm-cov --text"
coverage-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov.info"
coverage-json = "llvm-cov --json --output-path target/llvm-cov/coverage.json"
coverage-clean = "llvm-cov clean"

# Branch coverage commands (requires nightly Rust)
coverage-branch = "llvm-cov --html --open --branch"
coverage-branch-text = "llvm-cov --text --branch"
coverage-branch-lcov = "llvm-cov --lcov --output-path target/llvm-cov/lcov-branch.info --branch"
