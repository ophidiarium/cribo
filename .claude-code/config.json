{
    "hooks": {
        "preToolUse": "./.claude-code/hooks/prevent-main-commit.sh"
    }
}
