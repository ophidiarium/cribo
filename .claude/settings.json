{
    "$schema": "https://json.schemastore.org/claude-code-settings.json",
    "env": {
        "DISABLE_TELEMETRY": "1",
        "CARGO_TERM_COLOR": "always",
        "CLICOLOR": "1",
        "CLICOLOR_FORCE": "1",
        "NEXTEST_COLOR": "1",
        "CLAUDE_BASH_MAINTAIN_PROJECT_WORKING_DIR": "1",
        "GH_PAGER": "cat"
    },
    "permissions": {
        "allow": [
            "mcp__*",
            "Bash(gsed:*)",
            "Bash(git:*)",
            "Bash(git add:*)",
            "Bash(git commit:*)",
            "Bash(gh:*)",
            "Bash(grep:*)",
            "Bash(cargo:*)",
            "Bash(RUST_LOG=debug cargo:*)",
            "Bash(kill:*)",
            "Bash(curl:*)",
            "Bash(wget:*)",
            "Bash(cp:*)",
            "Bash(rustc:*)",
            "Bash(timeout:*)",
            "Bash(rg:*)",
            "Bash(sed:*)",
            "Bash(mkdir:*)",
            "Bash(chmod:*)",
            "Bash(cargo build:*)",
            "Bash(mv:*)",
            "Bash(ls:*)",
            "WebFetch(domain:github.com)",
            "WebFetch(domain:raw.githubusercontent.com)",
            "WebFetch(domain:pypi.org)"
        ],
        "deny": []
    },
    "hooks": {
        "PreToolUse": [
            {
                "matcher": "Bash",
                "hooks": [
                    {
                        "type": "command",
                        "command": "$CLAUDE_PROJECT_DIR/.claude/hooks/prevent-main-commit.sh"
                    }
                ]
            }
        ],
        "PostToolUse": [
            {
                "matcher": "Edit|Write",
                "hooks": [
                    {
                        "type": "command",
                        "command": "$CLAUDE_PROJECT_DIR/.claude/hooks/format-python.sh"
                    }
                ]
            }
        ]
    },
    "enableAllProjectMcpServers": true
}
