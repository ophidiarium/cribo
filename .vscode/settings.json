{
    "rust-analyzer.check.extraArgs": [
        "--all-features"
    ],
    "rust-analyzer.check.command": "clippy",
    "[json][jsonc][yaml][github-actions-workflow][python][toml]": {
        "editor.defaultFormatter": "dprint.dprint",
        "editor.formatOnSave": true
    },
    "cSpell.enabled": false,
    "chat.agent.enabled": true,
    "chat.promptFiles": true,
    "chat.instructionsFilesLocations": {
        ".github/instructions": true,
    },
    "github.copilot.chat.agent.thinkingTool": true,
    "github.copilot.advanced": {
        "useLanguageServer": true
    },
    "chat.mcp.enabled": true,
    "mcp": {
        "inputs": [
            {
                "type": "promptString",
                "id": "github_token",
                "description": "GitHub Personal Access Token",
                "password": true
            }
        ],
        "servers": {
            "context7": {
                "command": "npx",
                "args": [
                    "-y",
                    "@upstash/context7-mcp"
                ]
            },
            "github": {
                "command": "docker",
                "args": [
                    "run",
                    "-i",
                    "--rm",
                    "-e",
                    "GITHUB_PERSONAL_ACCESS_TOKEN",
                    "ghcr.io/github/github-mcp-server"
                ],
                "env": {
                    "GITHUB_PERSONAL_ACCESS_TOKEN": "${input:github_token}"
                }
            }
        }
    }
}