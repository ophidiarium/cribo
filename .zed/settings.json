{
    "file_scan_exclusions": [
        "**/.git",
        "**/Thumbs.dn",
        "**/.vscode",
        "**/__pycache__",
        "**/.mypy_cache",
        "**/.pytest_cache",
        "**/.ruff_cache",
        "**/.ropeproject"
    ],
    "auto_install_extensions": {
        "toml": true,
        "typos": true,
        "ruff": true,
        "mcp-server-github": true,
        "mcp-server-context7": true
    },
    "terminal": {
        "detect_venv": {
            "on": {
                "directories": [".venv"]
            }
        }
    }
}
