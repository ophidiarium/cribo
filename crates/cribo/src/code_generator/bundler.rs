use std::path::{Path, PathBuf};

use anyhow::{Context, Result, bail};
use indexmap::{IndexMap as FxIndexMap, IndexSet as FxIndexSet};
use log::{debug, trace, warn};
use ruff_python_ast::{
    self as ast, Arguments, CmpOp, Expr, ExprAttribute, ExprCall, ExprCompare, ExprContext,
    ExprDict, ExprList, ExprName, ExprStringLiteral, ExprSubscript, ExprTuple, Identifier, Int,
    Keyword, ModModule, Stmt, StmtAnnAssign, StmtAssign, StmtClassDef, StmtExpr, StmtFunctionDef,
    StmtIf, StmtImport, StmtImportFrom, StmtReturn, StmtWith, WithItem, visitor::Visitor,
};
use ruff_text_size::{Ranged, TextRange, TextSize};
use rustc_hash::{FxHashMap, FxHashSet};

use crate::{
    code_generator::{
        circular_deps::SymbolDependencyGraph,
        context::{
            BundleParams, DirectImportContext, HardDependency, InlineContext,
            ModuleTransformContext, ProcessGlobalsParams, SemanticContext, TransformFunctionParams,
        },
        globals::{GlobalsLifter, transform_globals_in_expr, transform_globals_in_stmt},
        import_transformer::{RecursiveImportTransformer, RecursiveImportTransformerParams},
    },
    cribo_graph::CriboGraph as DependencyGraph,
    semantic_bundler::{ModuleGlobalInfo, SemanticBundler, SymbolRegistry},
    transformation_context::TransformationContext,
    visitors::{ImportDiscoveryVisitor, NoOpsRemovalTransformer},
};

/// This approach avoids forward reference issues while maintaining Python module semantics
#[derive(Debug)]
pub struct HybridStaticBundler<'a> {
    /// Track if importlib was fully transformed and should be removed
    pub(crate) importlib_fully_transformed: bool,
    /// Map from original module name to synthetic module name
    pub(crate) module_registry: FxIndexMap<String, String>,
    /// Map from synthetic module name to init function name
    pub(crate) init_functions: FxIndexMap<String, String>,
    /// Collected future imports
    pub(crate) future_imports: FxIndexSet<String>,
    /// Collected stdlib imports that are safe to hoist
    /// Maps module name to map of imported names to their aliases (None if no alias)
    pub(crate) stdlib_import_from_map: FxIndexMap<String, FxIndexMap<String, Option<String>>>,
    /// Regular import statements (import module)
    pub(crate) stdlib_import_statements: Vec<Stmt>,
    /// Track which modules have been bundled
    pub(crate) bundled_modules: FxIndexSet<String>,
    /// Modules that were inlined (not wrapper modules)
    pub(crate) inlined_modules: FxIndexSet<String>,
    /// Entry point path for calculating relative paths
    pub(crate) entry_path: Option<String>,
    /// Entry module name
    pub(crate) entry_module_name: String,
    /// Whether the entry is __init__.py or __main__.py
    pub(crate) entry_is_package_init_or_main: bool,
    /// Module export information (for __all__ handling)
    pub(crate) module_exports: FxIndexMap<String, Option<Vec<String>>>,
    /// Lifted global declarations to add at module top level
    pub(crate) lifted_global_declarations: Vec<Stmt>,
    /// Modules that are imported as namespaces (e.g., from package import module)
    /// Maps module name to set of importing modules
    pub(crate) namespace_imported_modules: FxIndexMap<String, FxIndexSet<String>>,
    /// Reference to the central module registry
    pub(crate) module_info_registry: Option<&'a crate::orchestrator::ModuleRegistry>,
    /// Modules that are part of circular dependencies
    pub(crate) circular_modules: FxIndexSet<String>,
    /// Pre-declared symbols for circular modules (module -> symbol -> renamed)
    pub(crate) circular_predeclarations: FxIndexMap<String, FxIndexMap<String, String>>,
    /// Hard dependencies that need to be hoisted
    pub(crate) hard_dependencies: Vec<HardDependency>,
    /// Symbol dependency graph for circular modules
    pub(crate) symbol_dep_graph: SymbolDependencyGraph,
    /// Module ASTs for resolving re-exports
    pub(crate) module_asts: Option<Vec<(String, ModModule, PathBuf, String)>>,
    /// Global registry of deferred imports to prevent duplication
    /// Maps (module_name, symbol_name) to the source module that deferred it
    pub(crate) global_deferred_imports: FxIndexMap<(String, String), String>,
    /// Track all namespaces that need to be created before module initialization
    /// This ensures parent namespaces exist before any submodule assignments
    pub(crate) required_namespaces: FxIndexSet<String>,
    /// Runtime tracking of all created namespaces to prevent duplicates
    /// This includes both pre-identified and dynamically created namespaces
    pub(crate) created_namespaces: FxIndexSet<String>,
    /// Modules that have explicit __all__ defined
    pub(crate) modules_with_explicit_all: FxIndexSet<String>,
    /// Transformation context for tracking node mappings
    pub(crate) transformation_context: TransformationContext,
    /// Module/symbol pairs that should be kept after tree shaking
    pub(crate) tree_shaking_keep_symbols: Option<indexmap::IndexSet<(String, String)>>,
    /// Whether to use the module cache model for circular dependencies
    pub(crate) use_module_cache_model: bool,
}

impl<'a> Default for HybridStaticBundler<'a> {
    fn default() -> Self {
        Self {
            importlib_fully_transformed: false,
            module_registry: FxIndexMap::default(),
            init_functions: FxIndexMap::default(),
            future_imports: FxIndexSet::default(),
            stdlib_import_from_map: FxIndexMap::default(),
            stdlib_import_statements: Vec::new(),
            bundled_modules: FxIndexSet::default(),
            inlined_modules: FxIndexSet::default(),
            entry_path: None,
            entry_module_name: String::new(),
            entry_is_package_init_or_main: false,
            module_exports: FxIndexMap::default(),
            lifted_global_declarations: Vec::new(),
            namespace_imported_modules: FxIndexMap::default(),
            module_info_registry: None,
            circular_modules: FxIndexSet::default(),
            circular_predeclarations: FxIndexMap::default(),
            hard_dependencies: Vec::new(),
            symbol_dep_graph: SymbolDependencyGraph::default(),
            module_asts: None,
            global_deferred_imports: FxIndexMap::default(),
            required_namespaces: FxIndexSet::default(),
            created_namespaces: FxIndexSet::default(),
            modules_with_explicit_all: FxIndexSet::default(),
            transformation_context: TransformationContext::new(),
            tree_shaking_keep_symbols: None,
            use_module_cache_model: true,
        }
    }
}

// Main implementation
impl<'a> HybridStaticBundler<'a> {
    /// Create a new bundler instance
    pub fn new() -> Self {
        Self::default()
    }

    /// Check if a statement is a hoisted import
    pub fn is_hoisted_import(&self, stmt: &Stmt) -> bool {
        match stmt {
            Stmt::Import(import) => {
                // Check if any alias is in our stdlib imports
                import.names.iter().any(|alias| {
                    self.stdlib_import_from_map
                        .contains_key(&alias.name.to_string())
                })
            }
            Stmt::ImportFrom(from_import) => {
                if let Some(module) = &from_import.module {
                    let module_name = module.to_string();
                    self.stdlib_import_from_map.contains_key(&module_name)
                } else {
                    false
                }
            }
            _ => false,
        }
    }

    /// Resolve a relative import with context
    pub fn resolve_relative_import_with_context(
        &self,
        current_module: &str,
        module_path: &Path,
        level: u32,
        module: Option<&str>,
    ) -> Option<String> {
        // Implementation from original file
        // This is a placeholder - the actual implementation is quite complex
        None
    }

    /// Filter exports based on tree shaking
    pub fn filter_exports_by_tree_shaking(
        &self,
        keep_symbols: Option<&indexmap::IndexSet<(String, String)>>,
        module_name: &str,
        exports: &[String],
    ) -> Vec<String> {
        if let Some(keep_symbols) = keep_symbols {
            exports
                .iter()
                .filter(|symbol| {
                    keep_symbols.contains(&(module_name.to_string(), symbol.to_string()))
                })
                .cloned()
                .collect()
        } else {
            exports.to_vec()
        }
    }

    // More methods to be moved from the original implementation...
}

/// Main entry point for bundling modules
pub fn bundle_modules(params: BundleParams) -> Result<ModModule> {
    // This function implementation is very large and should be moved from the original file
    // For now, creating a placeholder
    let mut bundler = HybridStaticBundler::new();

    // The actual implementation involves:
    // 1. Processing modules and collecting information
    // 2. Handling circular dependencies
    // 3. Generating the bundled output
    // 4. Managing imports and exports

    // Placeholder return
    Ok(ModModule {
        range: TextRange::default(),
        body: vec![],
        node_index: Default::default(),
    })
}

// Additional implementation blocks and helper functions should be moved here
// from the original code_generator.rs file
