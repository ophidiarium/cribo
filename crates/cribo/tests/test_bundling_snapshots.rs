#![allow(clippy::disallowed_methods)] // insta macros use unwrap internally

use anyhow::Result;
use std::fs;
use std::path::Path;
use std::process::Command;
use tempfile::TempDir;

use cribo::bundler::Bundler;
use cribo::config::Config;
use cribo::util::get_python_executable;

// Ruff linting integration for cross-validation
use ruff_linter::linter::{ParseSource, lint_only};
use ruff_linter::registry::Rule;
use ruff_linter::settings::{LinterSettings, flags};
use ruff_linter::source_kind::SourceKind;
use ruff_python_ast::PySourceType;

/// Structured execution results for better snapshot formatting
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct ExecutionResults {
    status: ExecutionStatus,
    stdout: String,
    stderr: String,
}

/// Sanitize file paths in error messages to make them deterministic for snapshots
fn sanitize_paths(text: &str) -> String {
    use regex::Regex;

    // Match common temporary directory patterns and replace with a generic placeholder
    let patterns = vec![
        // Unix/macOS temp paths like /var/folders/xyz/abc123/T/.tmpXYZ/file.py
        (r"/var/folders/[^/]+/[^/]+/T/\.[^/]+/", "<TMP>/"),
        // Standard Unix temp paths like /tmp/.tmpXYZ/file.py
        (r"/tmp/\.[^/]+/", "<TMP>/"),
        // Windows temp paths
        (
            r"C:\\Users\\[^\\]+\\AppData\\Local\\Temp\\[^\\]+\\",
            "<TMP>/",
        ),
        // Generic temp directory patterns
        (r"/[Tt]emp/[^/]+/", "<TMP>/"),
        // Python installation paths (varying versions and locations)
        (
            r"/opt/homebrew/Cellar/python@[\d.]+/[\d._]+/Frameworks/Python\.framework/Versions/[\d.]+/lib/python[\d.]+/",
            "<PYTHON_LIB>/",
        ),
        (r"/usr/lib/python[\d.]+/", "<PYTHON_LIB>/"),
        (r"C:\\Python\d+\\lib\\", "<PYTHON_LIB>/"),
        // Windows hosted tool cache paths (GitHub Actions)
        (
            r"C:\\hostedtoolcache\\windows\\Python\\[\d.]+\\x64\\Lib\\",
            "<PYTHON_LIB>/",
        ),
    ];

    let mut result = text.to_string();
    for (pattern, replacement) in patterns {
        let re = Regex::new(pattern).expect("Invalid regex pattern");
        result = re.replace_all(&result, replacement).to_string();
    }

    // Normalize Python error formatting differences between versions
    // Replace line numbers in importlib which vary between Python versions
    let importlib_line_re = Regex::new(r"line \d+, in import_module").expect("Invalid regex");
    result = importlib_line_re
        .replace_all(&result, "line <LINE>, in import_module")
        .to_string();

    // Remove lines that only contain caret/tilde indicators as they vary between Python versions
    let indicator_line_re = Regex::new(r"(?m)^\s+[\^~]+\s*\n").expect("Invalid regex");
    result = indicator_line_re.replace_all(&result, "").to_string();

    // Normalize Windows path separators in sanitized paths
    result = result.replace("<PYTHON_LIB>/importlib\\", "<PYTHON_LIB>/importlib/");

    result
}

#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
enum ExecutionStatus {
    Success,
    Failed(i32),
}

/// Ruff linting results for cross-validation
#[derive(Debug)]
#[allow(dead_code)] // Fields are used via Debug trait for snapshots
struct RuffLintResults {
    f401_violations: Vec<String>, // Unused imports
    f404_violations: Vec<String>, // Late future imports
    other_violations: Vec<String>,
    total_violations: usize,
}

/// Run ruff linting on bundled code to cross-validate import handling
fn run_ruff_lint_on_bundle(bundled_code: &str) -> RuffLintResults {
    // Create settings for multiple import-related rules with both F401 and F404 enabled
    let settings = LinterSettings {
        rules: [Rule::UnusedImport, Rule::LateFutureImport]
            .into_iter()
            .collect(),
        ..LinterSettings::default()
    };

    let path = Path::new("<bundled>.py");
    let source_kind = SourceKind::Python(bundled_code.to_string());

    let result = lint_only(
        path,
        None,
        &settings,
        flags::Noqa::Enabled,
        &source_kind,
        PySourceType::Python,
        ParseSource::None,
    );

    let mut f401_violations = Vec::new();
    let mut f404_violations = Vec::new();
    let mut other_violations = Vec::new();

    for message in &result.messages {
        if let Some(rule) = message.to_rule() {
            let location = message.compute_start_location();
            let violation_info = format!(
                "Line {}: {} - {}",
                location.line.get(),
                rule.noqa_code(),
                message.body()
            );

            match rule {
                Rule::UnusedImport => f401_violations.push(violation_info),
                Rule::LateFutureImport => f404_violations.push(violation_info),
                _ => other_violations.push(violation_info),
            }
        }
    }

    let total_violations = f401_violations.len() + f404_violations.len() + other_violations.len();

    RuffLintResults {
        f401_violations,
        f404_violations,
        other_violations,
        total_violations,
    }
}

/// Generic test that processes all fixture directories in tests/fixtures/bundling
/// Each directory should contain a main.py entry point and will be bundled and executed
#[test]
fn test_all_bundling_fixtures() -> Result<()> {
    let fixtures_dir = Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("tests")
        .join("fixtures")
        .join("bundling");

    // Skip if bundling directory doesn't exist
    if !fixtures_dir.exists() {
        return Ok(());
    }

    // Read all subdirectories in the bundling fixtures directory
    let entries = fs::read_dir(&fixtures_dir)?;
    let mut fixture_names = Vec::new();

    for entry in entries {
        let entry = entry?;
        let path = entry.path();

        if path.is_dir() {
            if let Some(name) = path.file_name().and_then(|n| n.to_str()) {
                fixture_names.push(name.to_string());
            }
        }
    }

    // Sort for deterministic test order
    fixture_names.sort();

    // Process each fixture directory
    for fixture_name in fixture_names {
        test_single_bundling_fixture(&fixtures_dir, &fixture_name)?;
    }

    Ok(())
}

/// Test a single bundling fixture directory
fn test_single_bundling_fixture(fixtures_dir: &Path, fixture_name: &str) -> Result<()> {
    let fixture_path = fixtures_dir.join(fixture_name);
    let main_py_path = fixture_path.join("main.py");

    // Skip if main.py doesn't exist
    if !main_py_path.exists() {
        // Skip silently or use conditional verbose logging
        if std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!("Skipping {}: no main.py found", fixture_name);
        }
        return Ok(());
    }

    // Check if this is an expected failure fixture (xfail prefix)
    let expects_failure = fixture_name.starts_with("xfail_");

    // Create temporary directory for output
    let temp_dir = TempDir::new()?;
    let bundle_path = temp_dir.path().join("bundled.py");

    // Configure bundler
    let config = Config::default();
    let mut bundler = Bundler::new(config);

    // Bundle the fixture
    bundler.bundle(&main_py_path, &bundle_path, false)?;

    // Optionally validate Python syntax before execution
    let python_cmd = get_python_executable();
    let syntax_check = Command::new(&python_cmd)
        .args(["-m", "py_compile"])
        .arg(&bundle_path)
        .output();
    if let Ok(output) = syntax_check {
        if !output.status.success() && std::env::var("RUST_TEST_VERBOSE").is_ok() {
            eprintln!(
                "Warning: Bundled code has syntax errors for fixture {}",
                fixture_name
            );
            eprintln!("Stderr: {}", String::from_utf8_lossy(&output.stderr));
        }
    }

    // Read the bundled code and normalize line endings for cross-platform compatibility
    let bundled_code = fs::read_to_string(&bundle_path)?;

    // Run ruff linting for cross-validation of unused imports elimination
    let ruff_results = run_ruff_lint_on_bundle(&bundled_code);

    // Execute the bundled code with Python and capture output
    // Use Python executable with virtual environment support
    let python_cmd = get_python_executable();

    let python_output = Command::new(&python_cmd)
        .arg(&bundle_path)
        .current_dir(temp_dir.path()) // Set working directory for consistent execution
        .output()
        .map_err(|e| {
            anyhow::anyhow!(
                "Failed to execute Python: {} (command: {} {:?})",
                e,
                python_cmd,
                bundle_path
            )
        })?;

    // Note: For timeout support, we'd need a more complex solution or external crate
    // The standard library doesn't provide built-in timeout for process execution

    // Check for unexpected Python execution failures
    if !python_output.status.success() && !expects_failure {
        let stderr = String::from_utf8_lossy(&python_output.stderr);
        let stdout = String::from_utf8_lossy(&python_output.stdout);

        // This is an unexpected failure - fail the test explicitly
        return Err(anyhow::anyhow!(
            "Python execution failed unexpectedly for fixture '{}':\n\
            Exit code: {}\n\
            Stdout:\n{}\n\
            Stderr:\n{}\n\n\
            If this failure is expected, rename the fixture directory with 'xfail_' prefix.",
            fixture_name,
            python_output.status.code().unwrap_or(-1),
            stdout.trim(),
            stderr.trim()
        ));
    }

    // Create separate snapshots using insta's named snapshot feature
    insta::with_settings!({
        snapshot_suffix => fixture_name,
        omit_expression => true
    }, {
        // Snapshot the bundled code
        insta::assert_snapshot!("bundled_code", bundled_code);

        // Create structured execution results snapshot
        let execution_status = if python_output.status.success() {
            ExecutionStatus::Success
        } else {
            ExecutionStatus::Failed(python_output.status.code().unwrap_or(-1))
        };

        let execution_results = ExecutionResults {
            status: execution_status,
            stdout: String::from_utf8_lossy(&python_output.stdout)
                .trim()
                .replace("\r\n", "\n")
                .to_string(),
            stderr: sanitize_paths(
                &String::from_utf8_lossy(&python_output.stderr)
                    .trim()
                    .replace("\r\n", "\n")
            ),
        };

        insta::assert_debug_snapshot!("execution_results", execution_results);

        // Snapshot ruff linting results for cross-validation
        insta::assert_debug_snapshot!("ruff_lint_results", ruff_results);
    });

    Ok(())
}
