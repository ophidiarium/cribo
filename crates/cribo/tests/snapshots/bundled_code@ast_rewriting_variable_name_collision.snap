---
source: crates/cribo/tests/test_bundling_snapshots.rs
input_file: crates/cribo/tests/fixtures/bundling/ast_rewriting_variable_name_collision/main.py
---
import sys, types
def __cribo_init___cribo_35664f_constants():
    if '__cribo_35664f_constants' in sys.modules:
        return sys.modules['__cribo_35664f_constants']
    module = types.ModuleType('__cribo_35664f_constants')
    module.__file__ = '/Volumes/workplace/GitHub/ophidiarium/cribo/crates/cribo/tests/fixtures/bundling/ast_rewriting_variable_name_collision/constants.py'
    sys.modules['__cribo_35664f_constants'] = module
    sys.modules['constants'] = module
    """\nConstants module with global variables.\nThese will conflict with config.py when bundled.\n"""
    API_URL = "https://api.constants.com"
    module.API_URL = API_URL
    VERSION = "1.0.0"
    module.VERSION = VERSION
    DEFAULT_TIMEOUT = 30
    module.DEFAULT_TIMEOUT = DEFAULT_TIMEOUT
    MAX_CONNECTIONS = 100
    module.MAX_CONNECTIONS = MAX_CONNECTIONS
    RETRY_DELAY = 5
    module.RETRY_DELAY = RETRY_DELAY
    return module
def __cribo_init___cribo_2092b2_config():
    if '__cribo_2092b2_config' in sys.modules:
        return sys.modules['__cribo_2092b2_config']
    module = types.ModuleType('__cribo_2092b2_config')
    module.__file__ = '/Volumes/workplace/GitHub/ophidiarium/cribo/crates/cribo/tests/fixtures/bundling/ast_rewriting_variable_name_collision/config.py'
    sys.modules['__cribo_2092b2_config'] = module
    sys.modules['config'] = module
    """\nConfig module with global variables.\nThese will conflict with constants.py when bundled.\n"""
    API_URL = "https://api.config.com"
    module.API_URL = API_URL
    VERSION = "2.1.0"
    module.VERSION = VERSION
    MAX_RETRIES = 3
    module.MAX_RETRIES = MAX_RETRIES
    CACHE_SIZE = 1000
    module.CACHE_SIZE = CACHE_SIZE
    return module
__cribo_modules = {'constants': '__cribo_35664f_constants', 'config': '__cribo_2092b2_config'}
__cribo_init_functions = {'__cribo_35664f_constants': __cribo_init___cribo_35664f_constants, '__cribo_2092b2_config': __cribo_init___cribo_2092b2_config}
class CriboBundledFinder:

    def __init__(self, module_registry, init_functions):
        self.module_registry = module_registry
        self.init_functions = init_functions

    def find_spec(self, fullname, path, target=None):
        if fullname in self.module_registry:
            synthetic_name = self.module_registry[fullname]
            if synthetic_name not in sys.modules:
                init_func = self.init_functions.get(synthetic_name)
                if init_func:
                    init_func()
            import importlib.util
            return importlib.util.find_spec(synthetic_name)
        return None
sys.meta_path.insert(0, CriboBundledFinder(__cribo_modules, __cribo_init_functions))
__cribo_init___cribo_35664f_constants()
__cribo_init___cribo_2092b2_config()
"""\nVariable name collision test.\nBoth modules define variables with the same names.\n"""
API_URL = sys.modules['constants'].API_URL
VERSION = sys.modules['constants'].VERSION
DEFAULT_TIMEOUT = sys.modules['constants'].DEFAULT_TIMEOUT
CONFIG_API_URL = sys.modules['config'].API_URL
CONFIG_VERSION = sys.modules['config'].VERSION
MAX_RETRIES = sys.modules['config'].MAX_RETRIES
def main():
    print(f"Constants API URL: {API_URL}")
    print(f"Config API URL: {CONFIG_API_URL}")
    print(f"Constants Version: {VERSION}")
    print(f"Config Version: {CONFIG_VERSION}")
    print(f"Default Timeout: {DEFAULT_TIMEOUT}")
    print(f"Max Retries: {MAX_RETRIES}")
    return {"constants_api_url": API_URL, "config_api_url": CONFIG_API_URL, "constants_version": VERSION, "config_version": CONFIG_VERSION, "default_timeout": DEFAULT_TIMEOUT, "max_retries": MAX_RETRIES}
if __name__ == "__main__":
    result = main()
    print("Result:", result)
